From 7be4881884f9eb8cf22320e4479d7d7b16b723d5 Mon Sep 17 00:00:00 2001
From: Bogdan Roman <bogdan-gabriel.roman@nxp.com>
Date: Tue, 28 Feb 2023 12:04:28 +0200
Subject: [PATCH 1/9] examples: refactor examples build structure

Each example might have different needs for building and installing
their target binaries. Thus, each example has been moved in its
standalone directory, having an associated Makefile where its rules are
implemented. Since they follow the same pattern of rules (build, clean &
install), a common set of rules is automatically generated in
'examples/Makefile' for each example. This set calls for each example's
particular implementation of the rules.
The file 'examples/common.mk' contains common variables used for
building examples.

Issue: ALB-9734
Upstream-Status: Pending 

Signed-off-by: Bogdan Roman <bogdan-gabriel.roman@nxp.com>
Signed-off-by: Zhantao Tang <zhantao.tang@windriver.com>
---
 .gitignore                                    | 10 +--
 examples/Makefile                             | 79 ++++++-------------
 examples/common.mk                            | 43 ++++++++++
 examples/hse-encrypt/Makefile                 | 16 ++++
 examples/{ => hse-encrypt}/hse-encrypt.c      |  0
 examples/hse-secboot/Makefile                 | 17 ++++
 examples/{ => hse-secboot}/hse-secboot.c      |  0
 examples/{ => hse-secboot}/keys-config.h      |  0
 examples/hse-sysimg/Makefile                  | 16 ++++
 examples/{ => hse-sysimg}/hse-sysimg.c        |  0
 examples/pkcs-key-provision/Makefile          | 16 ++++
 .../pkcs-key-provision.c                      |  0
 examples/pkcs-keyop/Makefile                  | 19 +++++
 examples/{ => pkcs-keyop}/pkcs-keyop.c        |  0
 14 files changed, 155 insertions(+), 61 deletions(-)
 create mode 100644 examples/common.mk
 create mode 100644 examples/hse-encrypt/Makefile
 rename examples/{ => hse-encrypt}/hse-encrypt.c (100%)
 create mode 100644 examples/hse-secboot/Makefile
 rename examples/{ => hse-secboot}/hse-secboot.c (100%)
 rename examples/{ => hse-secboot}/keys-config.h (100%)
 create mode 100644 examples/hse-sysimg/Makefile
 rename examples/{ => hse-sysimg}/hse-sysimg.c (100%)
 create mode 100644 examples/pkcs-key-provision/Makefile
 rename examples/{ => pkcs-key-provision}/pkcs-key-provision.c (100%)
 create mode 100644 examples/pkcs-keyop/Makefile
 rename examples/{ => pkcs-keyop}/pkcs-keyop.c (100%)

diff --git a/.gitignore b/.gitignore
index fa1b8cc..e651679 100644
--- a/.gitignore
+++ b/.gitignore
@@ -1,8 +1,8 @@
-examples/pkcs-keyop
-examples/pkcs-key-provision
-examples/hse-sysimg
-examples/hse-encrypt
-examples/hse-secboot
+examples/pkcs-keyop/pkcs-keyop
+examples/pkcs-key-provision/pkcs-key-provision
+examples/hse-sysimg/hse-sysimg
+examples/hse-encrypt/hse-encrypt
+examples/hse-secboot/hse-secboot
 *.so*
 libhse/obj
 libpkcs/obj
diff --git a/examples/Makefile b/examples/Makefile
index 663e984..edd3317 100644
--- a/examples/Makefile
+++ b/examples/Makefile
@@ -1,65 +1,32 @@
 #
-# Copyright 2021 NXP
+# Copyright 2021, 2023 NXP
 #
 
-CFLAGS ?= -Wall -g
-LDFLAGS ?= -lcrypto -lp11
+examples :=  pkcs-keyop pkcs-key-provision hse-encrypt hse-sysimg hse-secboot
+clean_examples = $(addprefix clean_,$(examples))
+install_examples = $(addprefix install_,$(examples))
 
-ifeq (,$(CROSS_COMPILE))
-    $(error CROSS_COMPILE is not set)
-endif
+.PHONY: all clean $(examples) $(clean_examples)
 
-ifeq (,$(OPENSSL_DIR))
-$(warning Path to cross-compiled OpenSSL not defined, using default location)
-endif
-OPENSSL_DIR ?= $(HOME)/openssl-aarch64
+define generate_build_rule
+$(1):
+	make -C ./$(1)
+endef
 
-ifeq (,$(LIBP11_DIR))
-$(warning Path to cross-compiled libp11 not defined, using default location)
-endif
-LIBP11_DIR ?= $(HOME)/libp11-aarch64
+define generate_clean_rule
+clean_$(1):
+	make -C ./$(1) clean
+endef
 
-ifeq (,$(HSE_FWDIR))
-$(warning Path to HSE firmware package not defined, using default location)
-endif
-HSE_FWDIR ?= $(HOME)/HSE_FW_S32G2_0_1_0_0
+define generate_install_rule
+install_$(1):
+	make -C ./$(1) install
+endef
 
-LIBHSE_SRCDIR ?= ../libhse
-LIBHSE_DIR ?= ..
+all: $(examples)
+clean: $(clean_examples)
+install: $(install_examples)
 
-LIBPKCS_SRCDIR ?= ../libpkcs
-
-INCLUDE_KEYOP ?= -I$(OPENSSL_DIR)/include \
-		 -I$(LIBP11_DIR)/include
-
-INCLUDE_LIBHSE := -I$(LIBHSE_SRCDIR) \
-		  -I$(HSE_FWDIR)/interface \
-		  -I$(HSE_FWDIR)/interface/config \
-		  -I$(HSE_FWDIR)/interface/inc_common \
-		  -I$(HSE_FWDIR)/interface/inc_services
-
-INCLUDE_LIBPKCS := -I$(OPENSSL_DIR)/include \
-		   -I$(LIBPKCS_SRCDIR)
-
-LIBS ?= -L$(OPENSSL_DIR)/lib \
-	-L$(LIBP11_DIR)/lib
-
-all: pkcs-keyop pkcs-key-provision hse-encrypt hse-sysimg hse-secboot
-
-pkcs-keyop: pkcs-keyop.c
-	$(CROSS_COMPILE)gcc $(LIBS) $(INCLUDE_KEYOP) $(CFLAGS) $^ -o $@ $(LDFLAGS)
-
-pkcs-key-provision: pkcs-key-provision.c
-	$(CROSS_COMPILE)gcc -L$(OPENSSL_DIR)/lib $(INCLUDE_LIBPKCS) $(CFLAGS) $^ -o $@ -lcrypto -ldl
-
-hse-encrypt: hse-encrypt.c
-	$(CROSS_COMPILE)gcc -L$(LIBHSE_DIR) $(INCLUDE_LIBHSE) $(CFLAGS) $^ -o $@ -lhse
-
-hse-sysimg: hse-sysimg.c
-	$(CROSS_COMPILE)gcc -L$(LIBHSE_DIR) $(INCLUDE_LIBHSE) $(CFLAGS) $^ -o $@ -lhse
-
-hse-secboot: hse-secboot.c
-	$(CROSS_COMPILE)gcc -L$(LIBHSE_DIR) -L$(OPENSSL_DIR)/lib -I$(OPENSSL_DIR)/include $(INCLUDE_LIBHSE) $(CFLAGS) $^ -o $@ -lcrypto -lhse
-
-clean:
-	rm -f pkcs-keyop pkcs-key-provision hse-encrypt hse-sysimg hse-secboot
+$(foreach example, $(examples), $(eval $(call generate_build_rule,$(example))))
+$(foreach example, $(examples), $(eval $(call generate_clean_rule,$(example))))
+$(foreach example, $(examples), $(eval $(call generate_install_rule,$(example))))
diff --git a/examples/common.mk b/examples/common.mk
new file mode 100644
index 0000000..39025ff
--- /dev/null
+++ b/examples/common.mk
@@ -0,0 +1,43 @@
+CFLAGS ?= -Wall -g
+LDFLAGS ?=
+
+ifeq (,$(CROSS_COMPILE))
+    $(error CROSS_COMPILE is not set)
+endif
+
+ifeq (,$(OPENSSL_DIR))
+$(warning Path to cross-compiled OpenSSL not defined, using default location)
+endif
+OPENSSL_DIR ?= $(HOME)/openssl-aarch64
+
+ifeq (,$(LIBP11_DIR))
+$(warning Path to cross-compiled libp11 not defined, using default location)
+endif
+LIBP11_DIR ?= $(HOME)/libp11-aarch64
+
+ifeq (,$(HSE_FWDIR))
+$(warning Path to HSE firmware package not defined, using default location)
+endif
+HSE_FWDIR ?= $(HOME)/HSE_FW_S32G2_0_1_0_0
+
+PKCS11HSE_DIR ?= $(HOME)/pkcs11-hse
+
+LIBHSE_SRCDIR ?= $(PKCS11HSE_DIR)/libhse
+LIBHSE_DIR ?= $(PKCS11HSE_DIR)
+LIBPKCS_SRCDIR ?= $(PKCS11HSE_DIR)/libpkcs
+
+INCLUDE_KEYOP ?= -I$(OPENSSL_DIR)/include \
+		 -I$(LIBP11_DIR)/include
+
+INCLUDE_LIBHSE := -I$(LIBHSE_SRCDIR) \
+		  -I$(HSE_FWDIR)/interface \
+		  -I$(HSE_FWDIR)/interface/config \
+		  -I$(HSE_FWDIR)/interface/inc_common \
+		  -I$(HSE_FWDIR)/interface/inc_services
+
+INCLUDE_LIBPKCS := -I$(OPENSSL_DIR)/include \
+		   -I$(LIBPKCS_SRCDIR)
+
+LD_OPENSSL := -L$(OPENSSL_DIR)/lib -lcrypto
+LD_LIBP11 := -L$(LIBP11_DIR)/lib -lp11
+LD_LIBHSE := -L$(LIBHSE_DIR) -lhse
diff --git a/examples/hse-encrypt/Makefile b/examples/hse-encrypt/Makefile
new file mode 100644
index 0000000..695cade
--- /dev/null
+++ b/examples/hse-encrypt/Makefile
@@ -0,0 +1,16 @@
+#
+# Copyright 2023 NXP
+#
+
+include ../common.mk
+
+all: hse-encrypt
+
+hse-encrypt: hse-encrypt.c
+	$(CROSS_COMPILE)gcc $(LDFLAGS) $(INCLUDE_LIBHSE) $(CFLAGS) $^ -o $@ $(LD_LIBHSE)
+
+clean:
+	rm -f hse-encrypt
+
+install:
+	install hse-encrypt $(EXAMPLES_INSTALLDIR)
\ No newline at end of file
diff --git a/examples/hse-encrypt.c b/examples/hse-encrypt/hse-encrypt.c
similarity index 100%
rename from examples/hse-encrypt.c
rename to examples/hse-encrypt/hse-encrypt.c
diff --git a/examples/hse-secboot/Makefile b/examples/hse-secboot/Makefile
new file mode 100644
index 0000000..e3e1b31
--- /dev/null
+++ b/examples/hse-secboot/Makefile
@@ -0,0 +1,17 @@
+#
+# Copyright 2023 NXP
+#
+
+include ../common.mk
+
+all: hse-secboot
+
+hse-secboot: hse-secboot.c
+	$(CROSS_COMPILE)gcc $(LDFLAGS) -I$(OPENSSL_DIR)/include $(INCLUDE_LIBHSE) $(CFLAGS) $^ -o $@ \
+	$(LD_OPENSSL) $(LD_LIBHSE)
+
+clean:
+	rm -f hse-secboot
+
+install:
+	install hse-secboot $(EXAMPLES_INSTALLDIR)
diff --git a/examples/hse-secboot.c b/examples/hse-secboot/hse-secboot.c
similarity index 100%
rename from examples/hse-secboot.c
rename to examples/hse-secboot/hse-secboot.c
diff --git a/examples/keys-config.h b/examples/hse-secboot/keys-config.h
similarity index 100%
rename from examples/keys-config.h
rename to examples/hse-secboot/keys-config.h
diff --git a/examples/hse-sysimg/Makefile b/examples/hse-sysimg/Makefile
new file mode 100644
index 0000000..3c36395
--- /dev/null
+++ b/examples/hse-sysimg/Makefile
@@ -0,0 +1,16 @@
+#
+# Copyright 2023 NXP
+#
+
+include ../common.mk
+
+all: hse-sysimg
+
+hse-sysimg: hse-sysimg.c
+	$(CROSS_COMPILE)gcc $(LDFLAGS) $(INCLUDE_LIBHSE) $(CFLAGS) $^ -o $@ $(LD_LIBHSE)
+
+clean:
+	rm -f hse-sysimg
+
+install:
+	install hse-sysimg $(EXAMPLES_INSTALLDIR)
\ No newline at end of file
diff --git a/examples/hse-sysimg.c b/examples/hse-sysimg/hse-sysimg.c
similarity index 100%
rename from examples/hse-sysimg.c
rename to examples/hse-sysimg/hse-sysimg.c
diff --git a/examples/pkcs-key-provision/Makefile b/examples/pkcs-key-provision/Makefile
new file mode 100644
index 0000000..daed62f
--- /dev/null
+++ b/examples/pkcs-key-provision/Makefile
@@ -0,0 +1,16 @@
+#
+# Copyright 2023 NXP
+#
+
+include ../common.mk
+
+all: pkcs-key-provision
+
+pkcs-key-provision: pkcs-key-provision.c
+	$(CROSS_COMPILE)gcc $(LDFLAGS) $(INCLUDE_LIBPKCS) $(CFLAGS) $^ -o $@ $(LD_OPENSSL) -ldl
+
+clean:
+	rm -f pkcs-key-provision
+
+install:
+	install pkcs-key-provision $(EXAMPLES_INSTALLDIR)
\ No newline at end of file
diff --git a/examples/pkcs-key-provision.c b/examples/pkcs-key-provision/pkcs-key-provision.c
similarity index 100%
rename from examples/pkcs-key-provision.c
rename to examples/pkcs-key-provision/pkcs-key-provision.c
diff --git a/examples/pkcs-keyop/Makefile b/examples/pkcs-keyop/Makefile
new file mode 100644
index 0000000..3a68f42
--- /dev/null
+++ b/examples/pkcs-keyop/Makefile
@@ -0,0 +1,19 @@
+#
+# Copyright 2023 NXP
+#
+
+include ../common.mk
+
+INCLUDE_KEYOP ?= -I$(OPENSSL_DIR)/include \
+		 -I$(LIBP11_DIR)/include
+
+all: pkcs-keyop
+
+pkcs-keyop: pkcs-keyop.c
+	$(CROSS_COMPILE)gcc $(LDFLAGS) $(INCLUDE_KEYOP) $(CFLAGS)  $^ -o $@ $(LD_OPENSSL) $(LD_LIBP11)
+
+clean:
+	rm -f pkcs-keyop
+
+install:
+	install pkcs-keyop $(EXAMPLES_INSTALLDIR)
\ No newline at end of file
diff --git a/examples/pkcs-keyop.c b/examples/pkcs-keyop/pkcs-keyop.c
similarity index 100%
rename from examples/pkcs-keyop.c
rename to examples/pkcs-keyop/pkcs-keyop.c
-- 
2.25.1

